libgalois/include/katana/ArrowRandomAccessBuilder.h

```diff
@@ -10,182 +11,186 @@ namespace katana {
 
 namespace {
 
-/// NoNullBuilder uses std::vector for storage
+/// VectorBackedBuilder uses std::vector for storage
 /// Finalize() makes a copy of the data
-/// Does not support null values
+/// Thread-safe for concurrent accesses to different indices
+/// - Guaranteed by null mask stored as bytes in std::vector
 template <typename ValueType, typename StorageType, typename ArrowType>
-class NoNullBuilder {
+class VectorBackedBuilder {
 public:
   using value_type = ValueType;
   using reference = ValueType&;
 
-  NoNullBuilder(size_t length) : data_(length) {
+  VectorBackedBuilder(size_t length) : data_(length), valid_(length, false) {
     static_assert(sizeof(ValueType) == sizeof(StorageType));
   }
 
-  reference operator[](size_t index) {
+  // NOTE this operator has side-effects. It can safely be used in two ways:
+  // 1) builder[index] = value; where it creates a non-null entry
+  // 2) value = builder[index]; ONLY IF option 1 has already used that index
```

Comment:
What if we just didn't offer this operator? Is it common for people to need to read out of a builder?

Also could you elaborate on this comment? For my education?

Reply:
Reading out of a builder is (or at least was) used in the partitioner in the master->mirrors stage.
This comment indicates that `builder[index] = value` is safe, and sets the value at index as valid. But `builder[index]` has to set that value to valid when it's called, before it returns the reference, and it doesn't know if that reference is actually assigned to. So a raw `builder[index];` line is valid c++, but would create a valid value (0/"") erroneously